### PR TITLE
feat: filesystem navigator

### PR DESCRIPTION
**Pull request by tomzohar**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: [CONTRIBUTING.md](CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?
add file-system navigator 

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #5 

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information

image

### CODE REVIEW

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.html [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    <ng-container *ngFor="let dir of directories$ | async">
+      <mat-list-item (click)="onDirClicked(dir)">
+        <mat-icon matListItemIcon>folder</mat-icon>
+
+        <div matListItemTitle>
+          {{ dir.name }}
+
+          <img
+            *ngIf="dir.isNG"
+            src="https://angular.io/assets/images/logos/angular/angular.svg"
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.html:17]: move to consts, also use the same img for the angular logo in the header

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.html [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            width="20"
+            height="20"
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.html:19]: use css class instead of attributes

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  );
+  private readonly workspaceManagerApiService = inject(
+    WorkspaceManagerApiService
+  );
+
+  @Input() path: string | null = null;
+
+  @Output() pathChange = new EventEmitter<string>();
+
+  private readonly _isAngularWorkspace$ = new BehaviorSubject<boolean>(false);
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts:50]: remove _

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  readonly isAngularWorkspace$ = this._isAngularWorkspace$.asObservable();
+  directories$: Observable<Directory[]> = of([]);
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts:52]: change to BehaviorSubject

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+  ngOnChanges(): void {
+    if (this.path) {
+      this.directories$ = this.workspaceManagerApiService.getDirectoriesInPath(
+        this.path
+      );
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts:58]: directories.next instead of re-assigning. this format of managing directories will result in memory leaks

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    this.directories$
+      .pipe(
+        take(1),
+        switchMap((directories: Directory[]) => {
+          const dirname =
+            this.fsNavigatorUtilsService.extractDirNameFromPath(path);
+          const directory = directories.find((dir) => dir.name === dirname);
+          return directory
+            ? of(directory)
+            : this.fsNavigatorUtilsService.getDirectoryFromParentPath(
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts:72]: I don't think this component should make HTTP requests, I think it should be "stupider", and the component that uses it should send it the directories with input

File: libs/workspace-manager/src/lib/features/filesystem-navigator/services/filesystem-navigator-utils.service.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    return this.workspaceManagerApiService
+      .getDirectoriesInPath(parentPath)
+      .pipe(map((directories) => directories.find((d) => d.name === dirname)));
+  }
+
+  getParentPath(path: string): string {
+    const pathParts = path.split('/');
+    return pathParts.slice(0, pathParts.length - 1).join('/');
+  }
+}
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/services/filesystem-navigator-utils.service.ts:34]: I don't think this service should exist, create a utils directory on the same level as features and data access, move the getParentPath and extractDirNameFromPath to a pure functions in this directory. 

I think this `getDirectoryFromParentPath` should come from the component that uses the filesystem-navigator

**Review by OmerGronich — CHANGES_REQUESTED** [review]
please move filesystem-navigator and toolbar to workspace-manager/src/lib/ui instead of features

and also filesystem-navigator-utils should either be deleted or at the very least moved to data-access, and the word "utils" should be removed from the name

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts (hunk 6 above)
Comment (Tcodrz) [reply libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts:72]: so this is the approach I started with, but then i thought - wouldn't it be better to use a simpler API, just provide a path and the fs-navigator will take care of the rest. 
don't forget we will have a lot of events to take care of like refresh, this means a lot of event handling for each consumer of this component.

File: libs/workspace-manager/src/lib/features/filesystem-navigator/services/filesystem-navigator-utils.service.ts (hunk 7 above)
Comment (Tcodrz) [reply libs/workspace-manager/src/lib/features/filesystem-navigator/services/filesystem-navigator-utils.service.ts:34]: I tend to agree but this is a very common use-case that will need to be handled every time the user clicks on one of the breadcrumbs (path part), so the implementation will always be the same on every consumer of the component.
it will need to get the directories of the  parent path and find the current directory

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts (hunk 4 above)
Comment (Tcodrz) [reply libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts:52]: this means we expose the 'next' method for any outside consumer, you're OK with that ?

Comment (OmerGronich) [reply libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts]: Why? You can do this: 

private directoriesSubject = new BehaviorSubject<Directory[]>([]);
directories$ = this.directoriesSubject.asObservable();

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts (hunk 6 above)
Comment (OmerGronich) [reply libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts]: @ronnetzer @agroupp thoughts?

File: libs/workspace-manager/src/lib/consts/angular-logo-img.const.ts [hunk 8]
```diff
@@ -0,0 +1,2 @@
+export const ANGULAR_LOGO_IMG_SRC =
```
Comment (ronnetzer) [inline libs/workspace-manager/src/lib/consts/angular-logo-img.const.ts:1]: lets just download it and use it as an asset :)

File: libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  getHomeDir(): Observable<string> {
+    return this.http.get(`${this.API_URL}/homedir`, {
+      responseType: 'text',
+    });
+  }
+
+  getDirectoriesInPath(path: string): Observable<Directory[]> {
+    const params = path ? `path=${path}` : null;
+    const url = `${this.API_URL}/dir?${params ? params : ''}`;
+    return this.http.get<Directory[]>(url);
```
Comment (ronnetzer) [inline libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts:22]:
```suggestion
    return this.http.get<Directory[]>(`${this.API_URL}/dir`, { params: { path } });
```

File: libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class WorkspaceManagerApiService {
+  private readonly http = inject(HttpClient);
+  private readonly API_URL = '/api/workspace-manager';
+
+  getHomeDir(): Observable<string> {
+    return this.http.get(`${this.API_URL}/homedir`, {
+      responseType: 'text',
```
Comment (ronnetzer) [inline libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts:15]: it should also work without specifying this no?

File: libs/shared/data/src/lib/directory.interface.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
 export interface Directory {
   name: string;
-  isNG?: boolean;
+  isNG: boolean;
```
Comment (ronnetzer) [inline libs/shared/data/src/lib/directory.interface.ts:3]: can't it remain optional? why is this change needed

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+})
+export class FilesystemNavigatorToolbarComponent implements OnChanges {
+  @Input() path: string | null = null;
+  @Output() pathChange = new EventEmitter<string>();
+
+  pathParts: string[] = [];
+
+  ngOnChanges(): void {
+    if (typeof this.path === 'string') {
+      this.pathParts = this.path.split('/').filter((v) => !!v);
```
Comment (ronnetzer) [inline libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.ts:28]: you should make sure it works correctly for windows as well, maybe even add a dedicated endpoint that returns the separator of the current platform (its a built-in in nodejs)

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts (hunk 6 above)
Comment (ronnetzer) [reply libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator.component.ts]: Currently it's kinda both dumb and smart component, in general choose one of them and go with it. 
In this case I agree with @OmerGronich, lets make it dumb + service: don't write the logic of it in the consumer (like refresh and any other event) so it'll be easier to reuse it with other consumers

File: libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.ts (hunk 12 above)
Comment (Tcodrz) [reply libs/workspace-manager/src/lib/features/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.ts:28]: will look into it

File: libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts (hunk 10 above)
Comment (Tcodrz) [reply libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts:15]: since the return type is string i get an error because the 'regular' .get overload expects a JSON format.
we can either solve it like that in the FE or adjust the api to return `{path: string}`

**Comment by ronnetzer** [comment]
regarding the visuals of it, the toolbar should look like a filesystem path, you should change the`>` separator to per platform separator.
also note the angular icon looks like its in low resolution, also its not aligned with the text. (personally I think it should replace the directory icon or it should be moved to the end of the row [right side])

File: libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts (hunk 10 above)
Comment (Tcodrz) [reply libs/workspace-manager/src/lib/data-access/workspace-manager-api.service.ts:15]: image

**Comment by Tcodrz** [comment]
> regarding the visuals of it, the toolbar should look like a filesystem path, you should change the`>` separator to per platform separator. also note the angular icon looks like its in low resolution, also its not aligned with the text. (personally I think it should replace the directory icon or it should be moved to the end of the row [right side])

@OmerGronich WDYT about replacing the folder icon with the angular logo instead of adding the angular logo at the end ?
image

**Comment by OmerGronich** [comment]
> > regarding the visuals of it, the toolbar should look like a filesystem path, you should change the`>` separator to per platform separator. also note the angular icon looks like its in low resolution, also its not aligned with the text. (personally I think it should replace the directory icon or it should be moved to the end of the row [right side])
> 
> @OmerGronich WDYT about replacing the folder icon with the angular logo instead of adding the angular logo at the end ? image

Not bad IMO, @Danieliverant what do you think?

**Comment by Danieliverant** [comment]
> > > regarding the visuals of it, the toolbar should look like a filesystem path, you should change the`>` separator to per platform separator. also note the angular icon looks like its in low resolution, also its not aligned with the text. (personally I think it should replace the directory icon or it should be moved to the end of the row [right side])
> > 
> > 
> > @OmerGronich WDYT about replacing the folder icon with the angular logo instead of adding the angular logo at the end ? image
> 
> Not bad IMO, @Danieliverant what do you think?

I really like it

File: libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    component.path = '/mock/path';
+    component.separator = '/';
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  describe('ngOnChanges', () => {
```
Comment (ronnetzer) [inline libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.spec.ts:25]: most of the times you should stick to one top-level describe

File: libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  pathParts: string[] = [];
+
+  ngOnChanges(): void {
+    this.pathParts = (this.path as string)
+      .split(this.separator as string)
+      .filter((v) => !!v);
+  }
+
+  onPartClicked(pathPart: string): void {
+    const index = this.pathParts.findIndex((part) => part === pathPart);
```
Comment (ronnetzer) [inline libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.ts:34]:
```suggestion
    const index = this.pathParts.indexOf(pathPart);
```

File: libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    it('should split the path input and save the parts', () => {
+      component.ngOnChanges();
```
Comment (ronnetzer) [inline libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator-toolbar/filesystem-navigator-toolbar.component.spec.ts:27]: I think that if you'll use `fixture.componentRef.setInput(...)` instead of `component.` you won't need to manually trigger ngOnChanges

**Review by ronnetzer — APPROVED** [review]

File: libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+  mat-list-item:hover {
+    background: $list-item-hover;
+    cursor: pointer;
+  }
+
+  .angular-logo {
+    width: 25px;
+    height: 25px;
+  }
+}
```
Comment (OmerGronich) [inline libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss:20]: @Danieliverant Is this the way to use material colors in this scenario?

**Review by OmerGronich — APPROVED** [review]

File: libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss (hunk 16 above)
Comment (Danieliverant) [reply libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss:20]: no... why do you need to change the color? @tomzo42 
every material override should be done on the theme level not single component

Comment (Tcodrz) [reply libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss:20]: @Danieliverant I wanted to add a hover state for the list item, it doesn't have one by default

Comment (Danieliverant) [reply libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss:20]: @Tcodrz I don't think it's needed 
see material example:
https://material.angular.io/components/list/examples#list-sections

if you really want it you can create list.scss file in the theme, and override it there, if you want ping me on Slack we can do it together

Comment (Tcodrz) [reply libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss:20]: doesn't seem to have that hover effect like in the material example. 
IMO the list looks not so interactive without the hover, if I will use the theme for getting a grey color that would be OK or better to just leave it for now ?
@Danieliverant

Comment (Danieliverant) [reply libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss:20]: I guess it cuz you need to put `matListItemIcon` on the `mat-icon` itself and not on it's parent.

It's better to fix this ^ and leave it without a dedicated hover for now.
But if you want, you can override the hover behavior in the theme files

Comment (tomzohar) [reply libs/workspace-manager/src/lib/ui/filesystem-navigator/filesystem-navigator.component.scss]: @Danieliverant doesn't work either, I removed the css for now, will check it out in a different PR

**Review by Danieliverant — APPROVED** [review]
